test/e2e/kappcontroller/packageinstall_test.go

```diff
@@ -172,6 +173,165 @@ stringData:
+kind: Package
+metadata:
+  name: pkg.test.carvel.dev.%[1]s
+spec:
+`, pkgVer)
+	// too bad there's no tools for templating yaml.
+	if kcVerC != "" {
+		pkgYAML += fmt.Sprintf(`  kappControllerVersionSelection:
+    constraints: "%s"
+`, kcVerC)
+	}
+	if k8sVerC != "" {
+		pkgYAML += fmt.Sprintf(`  kubernetesVersionSelection:
+    constraints: "%s"
+`, k8sVerC)
+	}
+	pkgYAML += fmt.Sprintf(`  refName: pkg.test.carvel.dev
+  version: %[1]s
+  licenses:
+  - Apache 2.0
+  capactiyRequirementsDescription: "cpu: 1,RAM: 2, Disk: 3"
+  releaseNotes: |
+    - Introduce simple-app package
+  releasedAt: 2021-05-05T18:57:06Z
+  template:
+    spec:
+      fetch:
+      - imgpkgBundle:
+          image: k8slt/kctrl-example-pkg:v1.0.0
+      template:
+      - ytt: {}
+      - kbld:
+          paths:
+          - "-"
+          - ".imgpkg/images.yml"
+      deploy:
+      - kapp:
+          inspect: {}
+`, pkgVer)
+	return pkgYAML
+}
+
+func generatePkgiYAML(pkgiName, kcVerC, k8sVerC string, annotations []string) string {
+	installPkgYAML := generatePkgYAML(kcVerC, k8sVerC, "1.0.0") + fmt.Sprintf(`
+---
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: %[1]s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/packageinstalls
+`, pkgiName)
+	if len(annotations) > 0 {
+		for _, ann := range annotations {
+			installPkgYAML += fmt.Sprintf(`    "%s": ""`, ann)
+		}
+	}
+	installPkgYAML += `
+spec:
+  serviceAccountName: kappctrl-e2e-ns-sa
+  packageRef:
+    refName: pkg.test.carvel.dev
+    versionSelection:
+      constraints: ">0.0.0"
+  values:
+  - secretRef:
+      name: pkg-demo-values
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: pkg-demo-values
+stringData:
+  values.yml: |
+    hello_msg: "hi"
+`
+	return installPkgYAML
+}
+
+func Test_PackageInstalled_FromPackageInstall_VersionConstraints(t *testing.T) {
+	const (
+		kubernetesVersionOverrideAnnotation     = "packaging.carvel.dev/ignore-kubernetes-version-selection"
+		kappControllerVersionOverrideAnnotation = "packaging.carvel.dev/ignore-kapp-controller-version-selection"
+	)
+
+	env := e2e.BuildEnv(t)
+	logger := e2e.Logger{}
+	kapp := e2e.Kapp{t, env.Namespace, logger}
+	sas := e2e.ServiceAccounts{env.Namespace}
+	name := "instl-pkg-test-version-constraints"
+
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("PackageInstall fails due to k8s version constraint", func() {
+		installPkgYAML := generatePkgiYAML(name, "", "<1.0.0", []string{}) + sas.ForNamespaceYAML()
+		out, err := kapp.RunWithOpts([]string{"deploy", "-a", name, "-f", "-"},
+			e2e.RunOpts{
+				StdinReader: strings.NewReader(installPkgYAML),
+				AllowError:  true})
+		assert.Contains(t, out, "after-kubernetes-version-check=0")
+		require.Error(t, err)
+		// Without calling cleanup after each stage, somehow it seems like the controller doesn't
+		// get the updated annotation.
+		cleanUp()
```

Comment:
lets try to figure out since sounds like real users will run into this situation => update to annotation should "fix" the PackageInstall

Reply:
=> leave comment about metadata.generation not being incr when annotations change

Reply:
@cppforlife can we go over this one more time - still a little unclear to us exactly what is breaking and why it is/isn't a user-facing concern?